File: uilib/src/main/java/dev/entao/ui/util/WaveView.java
package dev.entao.ui.util;

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.view.View;
import dev.entao.appbase.App;

import java.util.ArrayList;
import java.util.List;
/**
 * Created by yangentao on 2015/11/22.
 * entaoyang@163.com
 */
public class WaveView extends View {
	private int maxVal = 0;
	private List<Integer> data = new ArrayList<>();
	private Paint paint = new Paint();

	public WaveView(Context context) {
		super(context);
		paint.setColor(Color.GRAY);
		paint.setStrokeWidth(App.INSTANCE.dp2px(2));
	}

	public void clearData(){
		data.clear();
		postInvalidate();
	}

	public WaveView setColor(int color) {
		paint.setColor(color);
		return this;
	}

	public WaveView setStrokeWidth(int dp) {
		paint.setStrokeWidth(App.INSTANCE.dp2px(dp));
		return this;
	}

	@Override
	protected void onDraw(Canvas canvas) {
		ArrayList<Integer> ls;
		synchronized (this) {
			ls = new ArrayList<>(data);
		}
		int size = ls.size();
		if (size == 0) {
			return;
		}
		int w = getWidth();
		int h = getHeight();
		double scaleH = h * 1.0 / maxVal;
		for (int i = 0; i < ls.size(); ++i) {
			ls.set(i, (int) (ls.get(i) * scaleH));
		}

		double f = w * 1.0 / size;//0.5
		int left = 0;
		for (int i = 0; i < size; ++i) {
			int val = ls.get(i);
			left = (int) (i * f);
			canvas.drawLine(left, h - val, left, h, paint);
		}

	}

	public WaveView setMaxValue(int max) {
		this.maxVal = max;
		return this;
	}

	public WaveView setValue(List<Integer> ls) {
		synchronized (this) {
			data = ls;
		}
		this.postInvalidate();
		return this;
	}

	public void show() {
		setVisibility(View.VISIBLE);
	}

	public void hide() {
		setVisibility(View.GONE);
	}
}

File: uilib/src/main/java/dev/entao/kan/util/WaveView.kt
package dev.entao.kan.util

import android.content.Context
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.view.View
import dev.entao.kan.appbase.App
import java.util.*

/**
 * Created by yangentao on 2015/11/22.
 * entaoyang@163.com
 */
class WaveView(context: Context) : View(context) {
    private var maxVal = 0
    private var data: MutableList<Int> = ArrayList()
    private val paint = Paint()

    init {
        paint.color = Color.GRAY
        paint.strokeWidth = App.dp2px(2).toFloat()
    }

    fun clearData() {
        data.clear()
        postInvalidate()
    }

    fun setColor(color: Int): WaveView {
        paint.color = color
        return this
    }

    fun setStrokeWidth(dp: Int): WaveView {
        paint.strokeWidth = App.dp2px(dp).toFloat()
        return this
    }

    override fun onDraw(canvas: Canvas) {
        val ls: ArrayList<Int>
        synchronized(this) {
            ls = ArrayList(data)
        }
        val size = ls.size
        if (size == 0) {
            return
        }
        val w = width
        val h = height
        val scaleH = h * 1.0 / maxVal
        for (i in ls.indices) {
            ls[i] = (ls[i] * scaleH).toInt()
        }

        val f = w * 1.0 / size//0.5
        var left = 0
        for (i in 0 until size) {
            val `val` = ls[i]
            left = (i * f).toInt()
            canvas.drawLine(left.toFloat(), (h - `val`).toFloat(), left.toFloat(), h.toFloat(), paint)
        }

    }

    fun setMaxValue(max: Int): WaveView {
        this.maxVal = max
        return this
    }

    fun setValue(ls: MutableList<Int>): WaveView {
        synchronized(this) {
            data = ls
        }
        this.postInvalidate()
        return this
    }

    fun show() {
        visibility = View.VISIBLE
    }

    fun hide() {
        visibility = View.GONE
    }
}
